executor/cluster_config.go

```diff
@@ -0,0 +1,198 @@
+// Copyright 2019 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package executor
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"sort"
+	"strings"
+	"sync"
+
+	"github.com/jeremywohl/flatten"
+	"github.com/pingcap/errors"
+	"github.com/pingcap/failpoint"
+	"github.com/pingcap/parser/terror"
+	"github.com/pingcap/tidb/infoschema"
+	plannercore "github.com/pingcap/tidb/planner/core"
+	"github.com/pingcap/tidb/types"
+	"github.com/pingcap/tidb/util"
+	"github.com/pingcap/tidb/util/chunk"
+	"github.com/pingcap/tidb/util/pdapi"
+)
+
+// ClusterConfigReader executes configuration retrieving from the cluster components
+type ClusterConfigReader struct {
```

Comment:
Can we define a more general `Reader` that has methods like `Next`, and all memory tables can implement such a `Reader`?
Because there will be more and more memory tables that filters can be pushed down. And it's redundant to write the same code for each memory table.

Reply:
Good idea, maybe I can refactor it after #13821 merged (If this PR is huge, prefer to refactor this in an individual PR).

Reply:
Can you split it into 2 PRs? One for add a general `Reader`, and one for reading cluster config.

Reply:
@djshow832 Good suggestion, I will split it into 2 PRs.

Reply:
@djshow832 Thanks for your advice, I have addressed it in the PR and it does not make this PR huge, so don't need to split it into 2 PRs.